src/config.c

```diff
@@ -2392,6 +2392,16 @@ static int isValidAnnouncedNodename(char *val,const char **err) {
     return 1;
 }
 
+static int isValidIpOrEmpty(char *val, const char **err) {
+    if (val[0] == '\0') {
+        return 1; /* empty is OK */
+    } else if (anetResolve(NULL, val, NULL, 0, ANET_IP_ONLY) == ANET_ERR) {
```

Comment:
Are there possible ways that you might announce an IPV6 address for a system were add only IPV4 is supported locally (perhaps in a container), where the resolve fails because it's not supported but is working today?

I like adding more validation, it just seems high risk because it might fail on upgrades. In the case you mentioned.

Reply:
Good question. AFACT from the man page of [getaddrinfo](https://www.man7.org/linux/man-pages/man3/getaddrinfo.3.html), this will accept IPv6 and IPv4 addresses, even if the system doesn't have IPv4 and IPv6 addresses configured, because we don't give the flag `AI_ADDRCONFIG`:

> If hints.ai_flags includes the AI_ADDRCONFIG flag, then IPv4 addresses are returned in the list pointed to by res only if the local system has at least one IPv4 address configured, and IPv6 addresses are returned only if the local system has at least one IPv6 address configured.  The loopback address is not  considered  for  this case as valid as a configured address.  This flag is useful on, for example, IPv4-only systems, to ensure that getaddrinfo() does not return IPv6 socket addresses that would always fail in connect(2) or bind(2).

I haven't tested though. Should I add an IPv6 address in the test so we cover that too?

> I like adding more validation, it just seems high risk because it might fail on upgrades. In the case you mentioned.

If it fails on upgrade because an invalid address was already configured, there was most likely already a problem. Or do you think any client accepts the form IP:port in the host field in CLUSTER SLOTS like that? A MOVED-redirect to this node with port or cluster-announce-port appended would contain the port twice, like `-MOVED 5555 10.1.150.58:6380:6380`, which doesn't make any sense.

Is it possible that anyone specifies a hostname for `cluster-announce-ip`? Would that work?

Reply:
Apparently, some users set a hostname for `cluster-announce-ip` so we must probably allow that too, if we don't want a breaking change. See this comment: https://github.com/redis/redis/pull/12581#issuecomment-1722532148

So @madolson, what do you think about a validation which allows IPv4, IPv6 or hostname? Or should we be even less strict?

Reply:
> If it fails on upgrade because an invalid address was already configured

We're assuming that it was an invalid address, but I really don't want to be missing something.

> So @madolson, what do you think about a validation which allows IPv4, IPv6 or hostname? Or should we be even less strict?

I'm not sure how to be less strict but still cover the case you mentioned. I think also adding the hostname check seems fine. We might also add a check though for the length, if someone is misusing the `cluster-announce-ip` we shouldn't be silently truncating the value we're storing.

Reply:
> I'm not sure how to be less strict but still cover the case you mentioned.

A regex which matches exactly an IPv4 address and a port `[0-9]+.[0-9]+.[0-9]+.[0-9]+:[0-9]+` covers only the case I mentioned :laughing: (though ridiculous).

I'll go for IPv4, IPv6 or hostname and a check for length < NET_IP_STR_LEN then.